### PR TITLE
Add cloning events from O1

### LINKED ISSUE #2773: Copying of content items?
Is there no facility for creating a new content item by copying an existing content item?  We can add/delete, but there's no copy feature that I can see.  This seems like a pretty fundamental feature that would be expected from any CMS.

### PR DESCRIPTION
**Pull request by jptissot**
Fixes #2773
Add cloning to content types.

### CODE REVIEW

**Comment by Skrypt** [comment]
Can you add the Clone button option in the OrchardCore.Contents views? That would be an easy way to test if it works. Add it to the button that has multiple options "Publish, Unpublish, Delete".

**Comment by jptissot** [comment]
I added the clone button. I tried cloning an item in the Admin and it did clone the Content Item (just didn't add any content but that is expected at this point I think.

**Comment by sebastienros** [comment]
I think it should actually clone the content in the Clone method. Then the event handlers give an opportunity for handlers bro change the value. Later in the Localization service there will be a Localize method that will call clone, and trigger Localizing events for handlers to react to it in case it's different than just cloning.

Try to "clone" the Content property and see what else needs to be done.

**Comment by jptissot** [comment]
I tried to clone the About article of the Blog recipe. It would not clone the Title for some reason. The rest seems to be cloned properly.

**Comment by jptissot** [comment]
I tested most fields and the bag part and they all seem to work properly.

When a content item has the list part however, the list contents do not get carried over.

**Comment by sebastienros** [comment]
Makes sense for the list part as the relationship is defined in the child item.

**Comment by jptissot** [comment]
What do you need me to do before this can be merged ? 

- Testing ?
- Fix ListPart ?

**Comment by sebastienros** [comment]
I won't merge it in dev before beta3 is released, it could add bugs that we don't want to focus on. I might merge it in a localization branch in the meantime if this can unblock you, and this way we can do other PR targeting this branch.

The List thing is worrying, maybe I should have designed it the same way as ContentPickerField. Maybe right now we can't even have an element in two lists, which could be an issue. Don't worry about it then, it's orthogonal ... I will review the code, test it and merge if it's ok.

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/CloneContentContext.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿namespace OrchardCore.ContentManagement.Handlers
+{
+    public class CloneContentContext : ContentContextBase
+    {
+        public ContentItem CloneContentItem { get; set; }
+        public string FieldName { get; set; } // TODO: check if this is required... Copied from O1
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/CloneContentContext.cs]: Please find out why it's there, and if it's not used in O1 just remove it then.

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/ContentPartHandler.cs [hunk 2]
```diff
@@ -155,6 +155,21 @@ async Task IContentPartHandler.GetContentItemAspectAsync(ContentItemAspectContex
                 await GetContentItemAspectAsync(context, (TPart)part);
             }
         }
+        async Task IContentPartHandler.CloningAsync(CloneContentContext context, ContentPart part)
+        {
+            if (part is TPart)
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/ContentPartHandler.cs]:
```
if (part is TPart tpart)
{
  await CloningAsync(context, tpart);
}

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs [hunk 3]
```diff
@@ -462,5 +462,23 @@ public async Task DiscardDraftAsync(ContentItem contentItem)
...
+            await CreateAsync(cloneContentItem, VersionOptions.Draft);
+
+            var context = new CloneContentContext(contentItem, cloneContentItem);
+
+            await Handlers.InvokeAsync(async handler => await handler.CloningAsync(context), _logger);
+
+            context.CloneContentItem.Data = contentItem.Data.DeepClone() as JObject;
+            context.CloneContentItem.DisplayText = contentItem.DisplayText;
+
+            await Handlers.InvokeAsync(async handler => await handler.ClonedAsync(context), _logger);
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: Use `ReverseHandlers`

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs [hunk 4]
```diff
@@ -462,5 +462,23 @@ public async Task DiscardDraftAsync(ContentItem contentItem)
...
         }
+
+        public async Task<ContentItem> CloneAsync(ContentItem contentItem)
+        {
+            var cloneContentItem = await NewAsync(contentItem.ContentType);
+            await CreateAsync(cloneContentItem, VersionOptions.Draft);
+
+            var context = new CloneContentContext(contentItem, cloneContentItem);
+
+            await Handlers.InvokeAsync(async handler => await handler.CloningAsync(context), _logger);
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: Move after the item is cloned, right before `ClonedAsync`. The idea here is that we just want a way to provide two steps for events. There is nothing that can be done before it's cloned.

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/ContentPartHandler.cs (hunk 2 above)
Comment (jptissot) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/ContentPartHandler.cs]: done

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs (hunk 3 above)
Comment (jptissot) [reply src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: Done

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs (hunk 4 above)
Comment (jptissot) [reply src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: Done

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/CloneContentContext.cs (hunk 1 above)
Comment (jptissot) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/CloneContentContext.cs]: Removed it without any effect.

**Comment by jptissot** [comment]
I implemented the Alias and Autoroute cloning handlers. Autoroute already had `GenerateUniquePathAsync` method. I copied the same logic for the Alias

**Pull request merged**